diff --git a/kiji-mapreduce/src/main/java/org/kiji/mapreduce/framework/KijiTableInputJobBuilder.java b/kiji-mapreduce/src/main/java/org/kiji/mapreduce/framework/KijiTableInputJobBuilder.java
index a2d2d53..4a61114 100644
--- a/kiji-mapreduce/src/main/java/org/kiji/mapreduce/framework/KijiTableInputJobBuilder.java
+++ b/kiji-mapreduce/src/main/java/org/kiji/mapreduce/framework/KijiTableInputJobBuilder.java
@@ -165,7 +165,7 @@ public abstract class KijiTableInputJobBuilder<T extends KijiTableInputJobBuilde
 
     // Validate the Kiji data request against the current table layout:
     Preconditions.checkNotNull(mInputTableURI, "Input Kiji table was never set.");
-    final Kiji kiji = Kiji.Factory.open(mInputTableURI);
+    final Kiji kiji = Kiji.Factory.open(mInputTableURI, job.getConfiguration());
     try {
       final KijiTable table = kiji.openTable(mInputTableURI.getTable());
       try {
