diff --git a/kiji-mapreduce/src/main/java/org/kiji/mapreduce/framework/KijiTableInputJobBuilder.java b/kiji-mapreduce/src/main/java/org/kiji/mapreduce/framework/KijiTableInputJobBuilder.java
index 27bf4d6..062558e 100644
--- a/kiji-mapreduce/src/main/java/org/kiji/mapreduce/framework/KijiTableInputJobBuilder.java
+++ b/kiji-mapreduce/src/main/java/org/kiji/mapreduce/framework/KijiTableInputJobBuilder.java
@@ -139,7 +139,7 @@ public abstract class KijiTableInputJobBuilder<T extends KijiTableInputJobBuilde
     super.configureJob(job);
 
     // Validate the Kiji data request against the current table layout:
-    final Kiji kiji = Kiji.Factory.open(mInputTableURI);
+    final Kiji kiji = Kiji.Factory.open(mInputTableURI, job.getConfiguration());
     try {
       final KijiTable table = kiji.openTable(mInputTableURI.getTable());
       try {
